### ISSUE TITLE
🏛 Registering Custom `SpeckleBase` Derived Classes

### ISSUE THREAD

**Issue by izzylys** [specklesystems/speckle-py#50]
Currently, only the `/speckle/objects/` dir is searched for Speckle classes. This was for ease of development and presuming we would have a large number of generated Speckle Objects from `speckle-sharp` by now. As the Objects generation has been temporarily parked, there is a more pressing need to allow users to register their own Speckle Objects to be picked up by the parser.

Implementation for this is up for discussion. @gjedlicska has proposed a solution involving metaclass magic, though this may be blocked by our current use of the pydantic base. Pydantic was chosen for the code generation niceties, but as this is not being utilised at the moment we are considering dropping it and finding a different solution for code generation when we get to that.

**Comment by gjedlicska** [comment]
Thanks for the writeup, I'm going to prepare a tech proposal for this.

**Comment by izzylys** [comment]
Sounds great! Looking forward to discussing this further and opening it up to the community ✨

**Comment by gjedlicska** [comment]
Hey, I know this has gotten long, but its a heavy topic. These are basically all the pros and cons I could conjure up.
I've done it both ways before, but in this case I see a clear winner (spoilers) in the form of doing it the metaclass way.
I'll try to come up with a good POC to support my brainfart.  

## Problem definition

The most important goal is the end user (package consumer) convenience. It should be easy and trivial to build on top of speckle_py. Currently the object deserialization is not easy to use. This is about finding a solution that provides good usability and doesn't over complicate the implementation.

## Typed object deserialization in python

I won't go into detail with serialization, this topic is complex enough on its own, plus serialization is a lot simpler. That should be handled separately.

In order to cast an incoming datastructure into a specific type object deserializers usually employ different tricks to achieve the desired result. In statically typed languages, usually a strongly defined object schema is used as a skeleton, that is filled with data by the parser. Some python implementations work similarly (pydantic), but things go sideways once generics or dynamic attributes come into play. Speckle rely heavily on dynamic attributes, so as far as my knowledge goes, there is no pre-baked solution that would solve the issue (if there is, pls let me know).

So the problem is two fold. Getting the intended type when parsing arbitrary models and handling dynamic attributes.

### Getting the type

The speckle transports convey each object in json format. These object have to be translatable to and from language native objects. This is basically a form of strongly typed json parsing. The speckle schema requires a unique type identifier for each type that is handled known as the `speckleType` or `speckle_type`.

Sidenote, pydantic can handle aliased parsing with snake_case to camelCase back and forth attribute transforms. This is quite handy to keep a pythonic object definition, yet still serialize into native javascript like attributes that are handled well in other parsers. (looking at you newtonsoft)

So the first problem to tackle is getting the proper `type` object inside python into which the incoming datastructure is parsed into (the type is initialized by recursive unpacking the incoming data dictionary using its keywords...). The python std lib json parser actually provides a mechanism to parse into specific types via the object_hook function. If the object hook function is capable of looking up the proper type, the de-serialization can be handled by the stdlib json parser.

The goal is to provide a suiting object_hook function.

I think the `speckle_type` attribute should be utilized to provide a mapping between the speckle_type attribute value and the actual python type object. This has a drawback, that no duplicate values are allowed. This has to be enforced by runtime guarantees. But its a tradeoff that I think is acceptable.

On to the question of registering `speckle_type` and type object mappings. In order to provide a proper mapping of types, There are two major ways to go about solving the problem. One is, that the `SpeckleBase` type takes care of registering any type, that subclasses it, into a register, or each class has to be semi-manually registered into a type mapping. The latter one is a lot simpler approach, but it requires the package users to conform into a certain way of registering their own custom types. This works in a closed environment, when all parties can be expected to know the rules. This trades implementation simplicity to user knowledge. The implementation uses `importlib`  to import specific types and register them in a type register. This becomes quickly cumbersome for anyone who just wants to build on top of speckle without knowing the inner implementation details. I think this is the majority of the speckle_py package users. So requiring package user cooperation should be minimized if there is a way.

The other way is relying on the Speckle base model to take care of registering each type that subclasses it. There is a valid way to go about this using a custom metaclass for the SpeckleBase model. A metaclass is basically a way how a type object is created in python and the basic mechanism can be built upon. Metaclass programming is considered as one of the black arts of python programming for a reason. It is not intuitive nor beginner friendly. Many crazy hacks can be done via metaclasses. This means, that the package and the maintainers take on the extra complexity for the sake of package user niceties. In this current case I think this is a worthy tradeoff to take. 

### Metaclassing

The way how the metaclass for speckle base should actually work is, that on each type definition, an entry is created into a registry (some well protected but global dict) with the actual `speckle_type`. This is done at parse time so there is an open question here, the speckle_type attribute is now a runtime value rather than a parse time. This means moving the type attribute value to a class attribute, but I think this should not cause any headaches. Based on the type value, the type creation code could actually register any subclassing type with its speckle type for parsing reuse. 

### Parsing

Given the global type registry it is quite straight forward to provide a function that can be used as the `object_hook`, that basically looks up speckle_type keys in the type registry.

This in turn results into a fully typed parsed model for nested datasctrucutes.

The performance implications should be tested...

### Extra attributes

The extra, non schema specific attributes should be handled by the base model. The basic python types have no issues with this, and pydantic can be configured to play nicely along.

### Pydantic and custom metaclasses

I've done some digging and pydantic does support adding custom metaclass behavior to base models, so this should be the way to go, to keep the doors open for easy code generation.

This whole topic has to be revised in the light of auto generated types, cause I'm not sure how that turns out. My gut feeling is, that it just works. but that has to be tested.

**Comment by gjedlicska** [comment]
@izzylys could you assign this to me?

**Comment by izzylys** [comment]
Thanks a lot for this writeup! I'll have a proper look over in the morning and am looking forward to the POC. I've just assigned the issue to you, so go right ahead with this! 🚀

**Comment by gjedlicska** [comment]
Ok, I've done some reseach, and I have to revise my shit. There is a 3rd way to do this since python 3.6 [PEP 487](https://www.python.org/dev/peps/pep-0487/#new-ways-of-using-classes), and it feels like a very solid solution. The code above may be ran as it is.

```python
from pydantic import BaseModel
from typing import ClassVar, Dict, Type


class ExampleBase(BaseModel):
    foo: int

    _registry : ClassVar[Dict[str, Type["ExampleBase"]]] = {}

    def __init_subclass__(cls, speckle_type:str, **kwargs):
        # this requires some validation, there is nothing blocking identical keys...
        cls.speckle_type = speckle_type
        ExampleBase._registry[speckle_type] = cls
        super().__init_subclass__(**kwargs)


class ExampleSub(ExampleBase, speckle_type="sub"):
    bar: str


class SpeckleSub(ExampleSub, speckle_type="speckle_sub"):
    magic : str


# # uncommenting this raises an error, for missing speckle_type definition
# # this automatically enforces mandatory type definition for everyone
# class FaultySub(ExampleBase):
    # pass


if __name__ == "__main__":

    for k,v in ExampleBase._registry.items():
        print(f"Speckle type {k}, mapped to {v.__name__}")

    print(ExampleBase._registry["speckle_sub"](foo=123, magic="trick", bar="baric").json())    ```

**Comment by gjedlicska** [comment]
One question here, I'm not sure about, is that this way, the `ExampleBase` class is not in registered in the mapping. 

This is means that there is no way to default back to parsing into a plain Base object. If that is desirable, it could be worked around with. Should this be supported @izzylys ?

**Comment by izzylys** [comment]
I like this! It seems much simpler and intuitive than the previously proposed solution. The one change I would make would be to not require the `speckle_type` and to populate that for them with `cls.__name__`. This is what is currently done in both `sharp` and `py` as that attribute is intended to be generated and not filled in by the user. 

On your question, deserialisation should definitely default to a regular `Base` if no matching `speckle_type` can be found. Similar to Kits in `sharp`, there is always a chance that a user doesn't have the objects or kits that the sender has. Failing to deserialise when the data is right there would be a very annoying fail.

**Comment by gjedlicska** [comment]
Sounds fair. Both are viable options. Docs should emphasize this both in class docstrings and user giude, that each class name has to be unique and if something is taken in the base kit, that that cannot be overriden. If that becomes an issue, it could be fixed by adding namespaces to the types, but that requires all speckle base kits to do the same. For now, lets go with this.
Thanks :)
